pkgs/misc/uboot/hardkernel-firmware.nix

```
@@ -0,0 +1,129 @@
+        buildPackages.gcc49
+      ] ++ lib.optional (stdenv.buildPlatform != stdenv.hostPlatform) buildPackages.stdenv.cc
+        ++ lib.optional (!stdenv.isAarch64) pkgsCross.aarch64-multiplatform.buildPackages.gcc49;
+
+      postPatch = ''
+        substituteInPlace Makefile --replace "/bin/pwd" "pwd"
+      '';
+
+      makeFlags = [
+        "CROSS_COMPILE=${stdenv.cc.targetPrefix}"
+        "CROSS_COMPILE_32=${pkgsCross.arm-embedded.stdenv.cc.targetPrefix}"
+        "${defconfig}" "bl301.bin"
+      ]
+      ++ lib.optional (!stdenv.isAarch64) "CROSS_COMPILE=${pkgsCross.aarch64-multiplatform.stdenv.cc.targetPrefix}";
+
+      installPhase = ''
+        mkdir -p ${installDir}
+        cp ${lib.concatStringsSep " " filesToInstall} ${installDir}
+      '';
+
+      meta = with lib; {
+        homepage = "https://www.hardkernel.com/";
+        description = "Das U-Boot from Hardkernel with Odroid embedded devices firmware and support";
+        license = licenses.unfreeRedistributableFirmware;
+        maintainers = with maintainers; [ aarapov ];
+      } // extraMeta;
+    } // removeAttrs args [ "extraMeta" ]);
+in {
+  inherit buildHardkernelFirmware;
+
+  # https://wiki.odroid.com/odroid-c2/software/building_u-boot
+  firmwareOdroidC2 = let
+    name = "odroidc2";
+    odroidc2-bl301 = fetchgit {
+      url = "https://github.com/hardkernel/u-boot_firmware.git";
+      rev = "b7b90c1099b057d35ebae886b7846b5d9bfb4143"; # "odroidc2-bl301"
+      sha256 = "0kdb1mg5zd7qyabfpbh98cs07icfzpkywvva13ybf4mf5g50g0n3";
+      deepClone = true;
+    };
+  in buildHardkernelFirmware {
+    defconfig = "odroidc2_config";
+    name = "firmware-odroid-c2";
+    version = "2015.01";
+    src = fetchgit {
+      url = "https://github.com/hardkernel/u-boot.git";
+      rev = "fac4d2da0a1b61dfdeaca0034a45151ff5983fb8"; # "odroidc2-v2015.01"
+      sha256 = "0jrpd7vww659nazyrv5af6n165akhz0h9hnxajq7gz906igc5raz";
+      leaveDotGit = true;
+    };
+
+    prePatch = ''
+      git remote add u-boot_firmware ${odroidc2-bl301}/
+      git fetch u-boot_firmware
+      git cherry-pick --no-commit \
+        5ce504067bb83de03d17173d5585e849df5d5a33^..${odroidc2-bl301.rev}
+
+      substituteInPlace ./arch/arm/cpu/armv8/gxb/firmware/scp_task/Makefile \
+        --replace "CROSS_COMPILE" "CROSS_COMPILE_32"
+    '';
+
+    filesToInstall = [
+      "build/scp_task/bl301.bin" "fip/gxb/bl30.bin" "fip/gxb/bl2.package"
+      "sd_fuse/sd_fusing.sh" "sd_fuse/bl1.bin.hardkernel"
+    ];
+
+    extraMeta.platforms = [ "aarch64-linux" ];
+  };
+
+  # https://wiki.odroid.com/odroid-c4/software/building_u-boot
+  firmwareOdroidC4 = buildHardkernelFirmware {
+    name = "firmware-odroid-c4";
+    defconfig = "odroidc4_defconfig";
+    version = "2015.01";
+    src = fetchgit {
+      url = "https://github.com/hardkernel/u-boot.git";
+      rev = "12c58e94e533b85a19d2f83d1c0a34345764ca07"; # "odroidg12-v2015.01"
+      sha256 = "06xmj8ksyx21qfdyq84lj029778vj30kwlrb91yj7bhwjxcz0qzn";
```

```suggestion
      sha256 = "110f8i2aj8p8b6ycw2hvilkp47vw754vdwv6vkpap8wmgkk90fjz";
```

That hash seems to be wrong:

    hash mismatch in fixed-output derivation '/nix/store/9mlfg5xlydnncrkin0l13vplf36ism82-u-boot-12c58e9':
      wanted: sha256:06xmj8ksyx21qfdyq84lj029778vj30kwlrb91yj7bhwjxcz0qzn
      got:    sha256:110f8i2aj8p8b6ycw2hvilkp47vw754vdwv6vkpap8wmgkk90fjz

---

Hmm... It is hardkernel's repo got a new commit and I thought I pinned it with `rev = ...;`.